File: 配套光盘/例题/0x60 图论/0x63 树的直径与最近公共祖先/树网的核/BZOJ1999.pas
{$M 100000000}
var
 ver,edge,next:array[0..1000000]of longint;
 head,d,f,q,a,fa:array[0..500000]of longint;
 v:array[0..500000]of boolean;
 n,s,m,tot,i,j,l,r,x,y,z,temp,ans:longint;

function min(x,y:longint):longint;
 begin
  if x<y then exit(x) else exit(y);
 end;
function max(x,y:longint):longint;
 begin
  if x>y then exit(x) else exit(y);
 end;

procedure add(x,y,z:longint);
 begin
  inc(tot);
  ver[tot]:=y;
  edge[tot]:=z;
  next[tot]:=head[x];
  head[x]:=tot;
 end;

procedure dfs(x:longint);
 var
  i:longint;
 begin
  i:=head[x];
  while i<>0 do
   begin
    if fa[x]<>ver[i] then
     begin
      fa[ver[i]]:=x;
      d[ver[i]]:=d[x]+edge[i];
      dfs(ver[i]);
     end;
    i:=next[i];
   end;
 end;

procedure diameter;
 begin
  dfs(1);
  x:=1;
  for i:=2 to n do
   if d[i]>d[x] then x:=i;
  d[x]:=0;
  fillchar(fa,sizeof(fa),0);
  dfs(x);
  y:=1;
  for i:=2 to n do
   if d[i]>d[y] then y:=i;
  while y<>x do
   begin
    v[y]:=true;
    inc(m); a[m]:=y;
    y:=fa[y];
   end;
  v[x]:=true;
  inc(m); a[m]:=x;
 end;

procedure treedp(x:longint);
 var
  i:longint;
 begin
  v[x]:=true;
  i:=head[x];
  while i<>0 do
   begin
    if not v[ver[i]] then
     begin
      treedp(ver[i]);
      f[x]:=max(f[x],f[ver[i]]+edge[i]);
     end;
    i:=next[i];
   end;
 end;

begin
 read(n,s);
 for i:=1 to n-1 do
  begin
   read(x,y,z);
   add(x,y,z);
   add(y,x,z);
  end;
 diameter;
 for i:=1 to m do treedp(a[i]);
 j:=m;
 ans:=maxlongint;
 temp:=0;
 for i:=1 to m do temp:=max(temp,f[a[i]]);
 for i:=m downto 1 do
  begin
   while (j>=1)and(d[a[j]]-d[a[i]]<=s) do dec(j);
   ans:=min(ans,max(temp,max(d[a[i]],d[a[1]]-d[a[j+1]])));
  end;
 writeln(ans);
end.

File: 配套光盘/例题/0x60 图论/0x63 树的直径与最近公共祖先/树网的核/BZOJ1999.cpp
#include <bits/stdc++.h>
using namespace std;
const int MAXN = 500005;
int ver[2*MAXN], nxt[2*MAXN], edge[2*MAXN], head[MAXN], tot;
int n, s, t;
queue<int> q;
int d[MAXN], pre[MAXN], a[MAXN], b[MAXN], f[MAXN], sum[MAXN];
bool v[MAXN];

void add(int x, int y, int z) {
    ver[++tot]=y, edge[tot]=z, nxt[tot]=head[x], head[x]=tot;
}

int bfs(int s) {
    memset(d, -1, sizeof(d));
    q.push(s); d[s] = 0;
    while (!q.empty()) {
        int x=q.front();
        q.pop();
        for (int i=head[x];i;i=nxt[i]) {
            int y=ver[i];
            if (d[y]==-1) {
                d[y]=d[x]+edge[i];
                pre[y]=i;
                q.push(y);
            }
        }
    }
    int p=s;
    for(int i=1;i<=n;i++)
        if(d[i]>d[p]) p=i;
    return p;
}

void dfs(int x) {
    v[x] = true;
    for (int i=head[x];i;i=nxt[i]) {
        int y=ver[i];
        if (v[y]) continue;
        dfs(y);
        f[x] = max(f[x], f[y] + edge[i]);
    }
}

int main() {
    cin >> n >> s;
    tot = 1;
    for (int i = 1; i < n; i++) {
        int x, y, z;
        scanf("%d%d%d", &x, &y, &z);
        add(x, y, z);
        add(y, x, z);
    }
    int p = bfs(1);
    int q = bfs(p);
    while (q!=p) {
        a[++t] = q;
        b[t+1] = edge[pre[q]];
        q = ver[pre[q]^1];
    }
    a[++t] = p;
    for (int i = 1; i <= t; i++) v[a[i]] = true;
    int maxf = 0;
    for (int i = 1; i <= t; i++) {
        dfs(a[i]);
        maxf = max(maxf, f[a[i]]);
        sum[i] = sum[i-1] + b[i];
    }
    int ans = 1 << 30;
    for (int i = 1, j = 1; i <= t; i++) {
        while (j < t && sum[j+1]-sum[i] <= s) j++;
        ans = min(ans, max(maxf, max(sum[i], sum[t] - sum[j])));
    }
    cout << ans << endl;
}